File: src/com/troystopera/gencode/exec/Scope.java
package com.troystopera.gencode.exec;

import com.troystopera.gencode.GenerationException;
import com.troystopera.gencode.code.components.Function;
import com.troystopera.gencode.var.ArrayVar;
import com.troystopera.gencode.var.Var;

import java.util.*;

/***
 * Representation of a scope. It maintains a list of all variables declared within it, as well as the values of
 * variables for use when executing the code to generate answers to questions.
 */
public class Scope {

    private static final Random random = new Random();

    //TODO: keep track of var types?
    private final Map<String, Var> vars = new HashMap<>();
    private final List<String> varNames = new ArrayList<>();
    private final Map<String, Function> funcs = new HashMap<>();
    private final List<String> funcNames = new ArrayList<>();
    private final Scope parent;
    private final int depth;

    private OutputListener listener;

    public Scope() {
        this(null, 0);
    }

    private Scope(Scope parent, int depth) {
        this.parent = parent;
        this.depth = depth;
    }

    public String getRandVarLimited() {
        return varNames.get(random.nextInt(varNames.size()));
    }

    public String getRandVar() {
        return getRandVar(depth);
    }

    public String getRandVar(int range) {
        return getParent(random.nextInt(range)).getRandVarLimited();
    }

    public boolean containsVarLimited(String var) {
        return varNames.contains(var);
    }

    public boolean containsVar(String var) {
        if (varNames.contains(var)) return true;
        else return hasParent() && getParent().containsVar(var);
    }

    public boolean containsFunction(String name) {
        return funcs.containsKey(name);
    }


    public boolean hasParent() {
        return parent != null;
    }

    public Scope getParent() {
        return parent;
    }

    public Scope getParent(int up) {
        Scope scope = this;
        for (int i = 0; i < up && scope.hasParent(); i++) {
            scope = scope.getParent();
        }
        return scope;
    }

    public int getScopeDepth() {
        return depth;
    }

    public void addVar(String name) {
        addVar(name, null);
    }

    public void addVar(String name, Var var) {
        varNames.add(name);
        vars.put(name, var);
    }

    public void nullVar(String var) {
        vars.remove(var);
    }

    public void nullArrVar(String var, int index) {
        ArrayVar array = (ArrayVar) getVal(var);
        if (array.getArray().length <= index)
            throw new GenerationException(new ArrayIndexOutOfBoundsException("Index " + index + " out of bounds"));
        array.getArray()[index] = null;
    }

    public Var getVal(String name) {
        Scope scope = this;
        while (!scope.vars.containsKey(name) && scope.hasParent()) scope = scope.getParent();
        if (scope.vars.containsKey(name)) return scope.vars.get(name);
        throw new GenerationException(new NullPointerException("Unknown variable: " + name));
    }

    public void updateVal(String name, Var var) {
        Scope scope = this;
        while (!scope.varNames.contains(name) && scope.hasParent()) scope = scope.getParent();
        if (scope.varNames.contains(name)) scope.vars.put(name, var);
        else throw new GenerationException(new NullPointerException("Unknown variable: " + name));
    }

    @SuppressWarnings("unchecked")
    public Var getArrVal(String name, int index) {
        ArrayVar<Var> array = (ArrayVar<Var>) getVal(name);
        if (array.getArray().length <= index)
            throw new GenerationException(new ArrayIndexOutOfBoundsException("Index " + index + " out of bounds"));
        return array.getArray()[index];
    }

    @SuppressWarnings("unchecked")
    public <T extends Var> void updateArrVal(String name, int index, T var) {
        ArrayVar<T> array = (ArrayVar<T>) getVal(name);
        if (array.getArray().length <= index)
            throw new GenerationException(new ArrayIndexOutOfBoundsException("Index " + index + " out of bounds"));
        array.getArray()[index] = var;
    }

    public Function getFunction(String name) {
        return funcs.get(name);
    }

    public void addFunction(String name, Function function) {
        funcNames.add(name);
        funcs.put(name, function);
    }

    public Scope newChildScope() {
        return new Scope(this, depth + 1);
    }

    public void setOutputListener(OutputListener listener) {
        this.listener = listener;
    }

    public interface OutputListener {

        void onWriteToConsole(String s);

        void onReturnValue(Var v);

    }

}

File: src/com/troystopera/gencode/exec/Scope.kt
package com.troystopera.gencode.exec

import com.troystopera.gencode.GenerationException
import com.troystopera.gencode.`var`.ArrayVar
import com.troystopera.gencode.`var`.Var
import com.troystopera.gencode.code.components.Function

class Scope protected constructor(val parent: Scope?, val depth: Int) {

    protected val functions = hashMapOf<String, Function>()
    protected val vars = hashMapOf<String, Var?>()

    constructor() : this(null, 0)

    fun getParent(up: Int = 1): Scope {
        var scope = this
        for (i in 1..up)
            if (scope.parent != null)
                scope = scope.getParent()
            else break
        return scope
    }

    fun hasVar(name: String): Boolean = vars.containsKey(name) || parent?.hasVar(name) ?: false

    fun addVar(name: String, v: Var? = null) = vars.put(name, v)

    fun getVar(name: String): Var {
        return vars[name] ?:
                parent?.getVar(name)
                ?: throw GenerationException(NoSuchElementException("Unknown variable: $name"))
    }

    fun setVar(name: String, v: Var) {
        if (vars.containsKey(name)) vars.put(name, v)
        else parent?.setVar(name, v) ?: throw GenerationException(NullPointerException("Unknown variable $name"))
    }

    @Suppress("UNCHECKED_CAST")
    fun getArrVar(name: String, index: Int): Var {
        val arr = getVar(name) as? ArrayVar<*> ?: throw GenerationException(IllegalArgumentException("Non array var: $name"))
        if (arr.array.size <= index) throw GenerationException(ArrayIndexOutOfBoundsException("Index $index out of bounds"))
        return arr.array[index]
    }

    @Suppress("UNCHECKED_CAST")
    fun <T : Var> setArrVar(name: String, index: Int, v: T) {
        val arr = getVar(name) as? ArrayVar<T> ?: throw GenerationException(IllegalArgumentException("Non array var: $name"))
        if (arr.array.size <= index) throw GenerationException(ArrayIndexOutOfBoundsException("Index $index out of bounds"))
        arr.array[index] = v
    }

    fun nullVar(name: String) = vars.put(name, null)

    @Suppress("UNCHECKED_CAST")
    fun nullArrVar(name: String, index: Int) {
        val arr = getVar(name) as? ArrayVar<Var> ?: throw GenerationException(IllegalArgumentException("Non array var: $name"))
        if (arr.array.size <= index) throw GenerationException(ArrayIndexOutOfBoundsException("Index $index out of bounds"))
        arr.array[index] = null
    }

    fun hasFun(name: String): Boolean = functions.containsKey(name) || (parent?.hasFun(name) ?: false)

    fun addFun(name: String, f: Function) = functions.put(name, f)

    fun getFun(name: String): Function = functions[name] ?: throw GenerationException(NoSuchElementException(name))

    open fun createChildScope(): Scope = Scope(this, depth + 1)

}